### DISCUSSION TITLE
Updating the iOS badge in a background push notification

### DISCUSSION THREAD

**Discussion by mark-kraemer** [excid3/noticed#397] [Ideas]
Assuming iOS notifications are in place, when a user reads a notification either in the app or from some other notification means, I would like to update badge for the iOS app. It's decently straight forward with Apnotic by just sending a push notification with no alert. Is there a way to do this with Noticed since it already handles the connection? If not, I can just utilize Apnotic directly. 

```ruby
    def update_badge(token)
      # Create a push notification for a specific device token
      background_notification = Apnotic::Notification.new(token)

      # without an alert attribute, this notification will be delivered in the background
      background_notification.badge = user.unread_notification_count
      background_notification.topic = APP_BUNDLE_IDENTIFIER

      # Send notification (this is a blocking call)
      connection.push(background_notification)
      # Close the connection
      connection.close
    end
```

**Comment by excid3** [comment]
You can do whatever you want with the notification: https://github.com/excid3/noticed/blob/main/lib/noticed/delivery_methods/ios.rb#L32-L47

**Reply by mark-kraemer** [reply]
That makes sense. After looking through the code, it doesn't look like there's an easy way to trigger a specific deliver method on a given notification. I could do something like the following that changes the apn.alert to nil when the notification is read. Did I miss an easier/better way to do this?

```ruby
notification.mark_as_read!
Noticed::DeliveryMethods::Ios.new.perform(:ios, notification)
```

**Reply by excid3** [reply]
I'm not following. Are you trying to update a notification after it's sent out?

**Reply by mark-kraemer** [reply]
Correct. Flow is the following:

1. Notification gets generated and sent out via the `deliver_by :ios`. The badge on the iOS app is updated to 1
2. User views the notification on desktop in browser and clicks on notification. This marks it as read
3. Another Apnotic notification is sent out in the background for that specific notification that only updates the badge

**Reply by excid3** [reply]
You can make another Notifier for that?

**Reply by mark-kraemer** [reply]
Ah, I see how that can work. 
1. An initial notification is created
2. When it is marked as read, a Noticed::Ephemeral notifier could trigger another notification that would handle the badge update.

**Reply by excid3** [reply]
I think that would probably work nicely. Let me know how it goes?

**Reply by mark-kraemer** [reply]
Will do, thanks for the suggestion!

**Reply by mark-kraemer** [reply]
@excid3 - Finally getting around to implementing this. I have the latest version (2.1.3) and I'm running into following error. Am I missing something?

 `<class:NativeBadgeNotifier>: undefined method notification_methods for NativeBadgeNotifier:Class (NoMethodError)`

```ruby
class NativeBadgeNotifier < Noticed::Ephemeral
  notification_methods do
  #   # Setting the alert text to nil will deliver the notification in
  #   # the background. This is used to update the app badge on the iOS home screen
    def ios_alert_text
      nil
    end
  end
end
```

**Reply by excid3** [reply]
Whoops, added it as an instance method instead of class method. Pushed that change to main.

**Reply by mark-kraemer** [reply]
Added some documentation on this pattern: https://github.com/excid3/noticed/pull/409.
